lldb/tools/lldb-server/lldb-platform.cpp

```diff
@@ -393,125 +362,222 @@ int main_platform(int argc, char *argv[]) {
-
-  do {
-    const bool children_inherit_accept_socket = true;
-    Connection *conn = nullptr;
-    error = acceptor_up->Accept(children_inherit_accept_socket, conn);
+  lldb_private::Args gdb_inferior_arguments;
+  std::unique_ptr<TCPSocket> sock_gdb;
+  if (protocol == Socket::ProtocolTcp) {
+    sock_gdb = std::make_unique<TCPSocket>(
+        /*should_close=*/true, /*child_processes_inherit=*/false);
+    error = sock_gdb->Listen(
+        llvm::formatv("{0}:{1}", hostname, gdbserver_port).str(), backlog);
     if (error.Fail()) {
-      WithColor::error() << error.AsCString() << '\n';
-      exit(socket_error);
+      printf("Failed to listen gdb: %s\n", error.AsCString());
+      return socket_error;
     }
-    printf("Connection established.\n");
+    if (gdbserver_port == 0)
+      gdbserver_port = sock_gdb->GetLocalPortNumber();
 
-    if (g_server) {
-      std::optional<uint16_t> available_port;
-      {
-        std::lock_guard<std::mutex> guard(gdbserver_portmap_mutex);
-        auto port = gdbserver_portmap.GetNextAvailablePort();
-        if (port)
-          available_port = *port;
-        else
-          llvm::consumeError(port.takeError());
-      }
-      if (!available_port) {
-        fprintf(stderr,
-                "no available gdbserver port for connection - dropping...\n");
-      } else {
-        const Socket *conn_socket =
-            static_cast<const Socket *>(conn->GetReadObject().get());
-        error =
-            spawn_process(progname, conn_socket, *available_port, port_offset,
-                          inferior_arguments, log_file, log_channels);
-        if (error.Fail()) {
-          {
-
-            std::lock_guard<std::mutex> guard(gdbserver_portmap_mutex);
-            gdbserver_portmap.FreePort(*available_port);
-          }
-          LLDB_LOGF(GetLog(LLDBLog::Platform), "spawn_process failed: %s",
-                    error.AsCString());
-          WithColor::error()
-              << "spawn_process failed: " << error.AsCString() << "\n";
-        }
-      }
-      // Parent doesn't need a connection to the lldb client
-      delete conn;
-
-      // Parent will continue to listen for new connections.
-      continue;
-    } else {
-      // If not running as a server, this process will not accept
-      // connections while a connection is active.
-      acceptor_up.reset();
-
-      // When not running in server mode, use all available ports
-      platform.SetPortMap(std::move(gdbserver_portmap));
+    gdb_inferior_arguments.AppendArguments(inferior_arguments);
+  }
+
+  GDBRemoteCommunicationServerPlatform platform(protocol, gdbserver_port);
+  {
+    llvm::Expected<std::vector<MainLoopBase::ReadHandleUP>> platform_handles =
+        sock_platform->Accept(
+            *g_main_loop,
+            [progname, gdbserver_port, inferior_arguments, log_file,
+             log_channels](std::unique_ptr<Socket> sock_up) {
+              // If not running as a server, this process will not accept
```

Comment:
AIUI, this is here to prevent multiple connections in non-server mode. Correct?

Is there any situation in non-server mode where we need to accept more than one successful connection?

Reply:
We must accept only 1 platform connection in non-server mode. But we can accept multiple gdb connections.

Reply:
Okay, in that case, I'd suggest that, instead of dropping connections on the floor, we close the listening handles (`platform_handles->clear()`) and also close the listening socket, so that the other side gets an ECONNREFUSED.

Reply:
In case of unix sockets (no gdb listener) it will cause to immediately exit the main loop, exit the app and kill the child process which handles the platform connection.

Reply:
I think that's actually great, as it allows us to synchronize with the exit of the child process and solve our global/detached thread problem -- after returning from MainLoop::Run, we just wait for a notification that the process has really exited. Something like:
```
std::promise<void> child_exited;
handles = socket.Accept(loop, [&]{
...
launch_info.SetMonitorProcessCallback([&child_exited]{child_exited.set_value();});
Launch(launch_info);
handles.clear();
});
loop.Run();
child_exited.get_future().get();
```

Reply:
Done.